### PR TITLE
[Enhancement] conditional expressions support array/map/struct/json types

### PR DESCRIPTION
**Pull request by fzhedu**
## Problem Summary:
Fixes # (issue)

conditional expressions support array/map/struct/json types, including `if`, `ifnull`, `nullif`, `coalesc`, `case... when...`

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 10 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|5|9|55.56%|[194, 195, 196, 197]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 10 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|5|9|55.56%|[194, 195, 196, 197]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 10 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|5|9|55.56%|[194, 195, 196, 197]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 10 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|5|9|55.56%|[194, 195, 196, 197]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|

File: be/src/exprs/condition_expr.cpp [hunk 1]
```diff
@@ -324,33 +400,72 @@ class VectorizedCoalesceExpr : public Expr {
...
-
-            // if last column
-            if (col == (col_size - 1)) {
-                if (viewers[col].is_null(row)) {
-                    builder.append_null();
-                } else {
-                    builder.append(viewers[col].value(row));
-                }
+            // if all nulls
+            if (col == col_size) {
```
Comment (Seaven) [inline be/src/exprs/condition_expr.cpp:431]: may append more value

File: be/src/exprs/condition_expr.cpp [hunk 2]
```diff
@@ -324,33 +400,72 @@ class VectorizedCoalesceExpr : public Expr {
...
+                builder.append_null();
             }
```
Comment (Seaven) [inline be/src/exprs/condition_expr.cpp:433]: may append more value

Comment (fzhedu) [reply be/src/exprs/condition_expr.cpp:433]: `if (col == col_size)` means all columns are null. if the last is not null,` col == col_size -1`.

**Review by Seaven — APPROVED** [review]

File: be/src/exprs/condition_expr.cpp [hunk 3]
```diff
@@ -111,6 +115,26 @@ class VectorizedIfNullExpr : public Expr {
...
+            null = down_cast<NullableColumn*>(columns[0].get())->null_column();
+        }
+
+        for (int row = 0; row < num_rows; ++row) {
+            if (null == nullptr || !null->get_data()[row]) { // not null
+                res->append(*columns[0], row, 1);
+            } else {
+                res->append(*columns[1], row, 1);
+            }
+        }
```
Comment (silverbullet233) [inline be/src/exprs/condition_expr.cpp:135]: in the case of `null == nullptr`, can we use a separate loop to append columns[0] directly without having to check inside each loop?

**Review by satanson — APPROVED** [review]

File: be/src/exprs/condition_expr.cpp (hunk 3 above)
Comment (fzhedu) [reply be/src/exprs/condition_expr.cpp:135]: branch prediction will optimize that case where a branch are almost true.

**Review by silverbullet233 — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**